staking/staking/staking_component.go

```diff
@@ -22,31 +28,222 @@ func NewStakingComponent(
+// The loop exits gracefully when the context is cancelled.
 func (sc *StakingComponentImpl) Start(ctx context.Context) error {
-	sc.logger.Debug("Starting Staking Component")
+	sc.logger.Debug("Starting Staking Component...")
 
-	err := sc.cardanoChainObserver.Start()
-	if err != nil {
-		return fmt.Errorf("failed to start observer for %s: %w", sc.cardanoChainObserver.GetConfig().GetChainID(), err)
+	// Start the Cardano chain observer
+	if err := sc.cardanoChainObserver.Start(); err != nil {
+		return fmt.Errorf("failed to start observer for chain %s: %w",
+			sc.cardanoChainObserver.GetConfig().GetChainID(), err)
 	}
 
-	waitTime := time.Millisecond * time.Duration(sc.config.PullTimeMilis)
+	waitInterval := time.Millisecond * time.Duration(sc.config.PullTimeMilis)
 
+	// Background loop that runs until context is cancelled
 	for {
 		select {
 		case <-ctx.Done():
+			sc.logger.Info("Staking Component shutting down...")
+
 			return nil
-		case <-time.After(waitTime):
+
+		case <-time.After(waitInterval):
+			sc.logger.Debug("Staking Component executing...")
+		}
+	}
+}
+
+// GetExchangeRate returns the current exchange rate between the native tokens and stTokens.
+func (sc *StakingComponentImpl) GetExchangeRate() float64 {
+	sc.mutex.RLock()
+	defer sc.mutex.RUnlock()
+
+	return sc.exchangeRate
+}
+
+// ChooseStakeAddrForStaking selects the staking address with the lowest current load.
+func (sc *StakingComponentImpl) ChooseStakeAddrForStaking(amount *big.Int) (string, error) {
+	if len(sc.stakingAddresses) == 0 {
+		return "", fmt.Errorf("no staking addresses configured for chainID %s", sc.config.Chain.ChainID)
+	}
+
+	minTotalTokens := new(big.Int)
+	addrForStaking := ""
+
+	for addr, stakeAddr := range sc.stakingAddresses {
+		totalTokens := stakeAddr.GetTotalTokensWithRewards()
+		if addrForStaking == "" || totalTokens.Cmp(minTotalTokens) < 0 {
+			minTotalTokens = totalTokens
+			addrForStaking = addr
+		}
+	}
+
+	return addrForStaking, nil
+}
+
+// ChooseStakeAddrForUnstaking selects the staking address with the most available tokens for unstaking.
+// Currently returns the address with the highest load.
+func (sc *StakingComponentImpl) ChooseStakeAddrForUnstaking(amount *big.Int) (map[string]*big.Int, error) {
+	if len(sc.stakingAddresses) == 0 {
+		return nil, fmt.Errorf("no staking addresses configured for chainID %s", sc.config.Chain.ChainID)
+	}
+
+	sorted := make([]core.StakingAddress, 0, len(sc.stakingAddresses))
+	for _, addr := range sc.stakingAddresses {
+		sorted = append(sorted, addr)
+	}
+
+	// Sort in descending order by total tokens with rewards
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i].GetTotalTokensWithRewards().Cmp(sorted[j].GetTotalTokensWithRewards()) > 0
+	})
+
+	remaining := new(big.Int).Set(amount)
+	addrsForUnstake := make(map[string]*big.Int)
+
+	for _, addr := range sorted {
+		if remaining.Cmp(big.NewInt(0)) == 0 || addr.GetTotalTokensWithRewards().Cmp(big.NewInt(0)) == 0 {
```

Comment:
just a suggestion: move `big.NewInt(0)` into some variable. Or even better use `BitLen() == 0` for zero comparasion or `Sign() > 0` for greater than zero comparasion